### ISSUE TITLE
Symlinking /var/lib/kubelet/pods (or /var/lib/kubelet) breaks the AWS EFS CSI Driver...

### ISSUE THREAD

**Issue by diranged** [bottlerocket-os/bottlerocket#2218]
**Image I'm using:**
Bottlerocket 1.8.0 (either AMD64 or ARM64)

**Background**
We recently had a conversation with @bcressey on Slack (https://awsdevelopers.slack.com/archives/CV5GQHN2Y/p1654555993450939) about mounting the "local disk" from EC2 instances such that Kubernetes would use them for container storage. The suggested answer at https://github.com/bottlerocket-os/bottlerocket/discussions/1991 was then modified based on some suggestions at https://gist.github.com/bcressey/2d3e139dbf01a1e1562f5ba49118258b, and we thought we had a working solution. 

**What went wrong?**

This morning at 2AM an engineering team escalated an issue to us - their Pods were failing to start up with errors like this:
```
Jun 15 08:35:45 ip-100-64-5-48.us-west-2.compute.internal containerd[10986]: time="2022-06-15T08:35:45.445512286Z" level=error msg="CreateContainer within sandbox \"7a431deadde10db2eb406da89caf5f77722f0841fd7c7e3497f7e70746575145\" for &ContainerMetadata{Name:taskmanager,Attempt:0,} failed" error="failed to generate container \"0c57cabdc198b1a68ed412a7f34dbe68b9004974d15f9146aa2b947c2f8d9e43\" spec: failed to generate spec: relabel \"/var/lib/kubelet/pods/e252e874-5b4e-4870-9fa5-633ff8eaa871/volumes/kubernetes.io~csi/pvc-be58bf32-bfaa-4518-9da0-3f9e41f85b21/mount\" with \"system_u:object_r:data_t:s0:c577,c1017\" failed: lsetxattr /var/lib/kubelet/pods/e252e874-5b4e-4870-9fa5-633ff8eaa871/volumes/kubernetes.io~csi/pvc-be58bf32-bfaa-4518-9da0-3f9e41f85b21/mount/55c0ed2511f6c836524254929bb885bc: operation not supported"
```

We discovered that the pods worked fine on hosts where our bootstrap container had not detected any ephemeral disks (`c6i`, `r6i`, etc..) - but they would fail 100% of the time when launching on a host where our bootstrap container had found local disks (`c6id`, etc).

**Root cause?**

While the examples in [discussion #1991](https://github.com/bottlerocket-os/bottlerocket/discussions/1991) and in @bcressey 's [gist](https://gist.github.com/bcressey/2d3e139dbf01a1e1562f5ba49118258b) technically work for most things.. they apparently break when interacting with the [aws-efs-csi-driver](https://github.com/kubernetes-sigs/aws-efs-csi-driver).

**Steps to reproduce?**

1. Create a bootstrap-container using https://gist.github.com/bcressey/2d3e139dbf01a1e1562f5ba49118258b
2. Create an EKS cluster with Bottlerocket 1.8.0, EKS 1.22, and install the aws-efs-csi-driver
4. Create an EFS PVC:
    ```yaml
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: test-pvc
    spec:
      accessModes:
      - ReadWriteMany
      resources:
        requests:
          storage: 10Gi
      storageClassName: efs
      volumeMode: Filesystem
     ```
5. Create a Pod mounting this PVC
6. Watch it fail

**Thoughts/Suggestions?**

I think it's time that Bottlerocket have an official way to manage storage _before_ kubelet starts up... or that there are parameters exposed that let us reconfigure kubelet to point to a different storage path. Either way, the symlink option always felt like a hack... and the hack finally bit us.

**Note on `/var/lib/kubelet` vs `/var/lib/kubelet/pods`**

We actually move and symlink `/var/lib/kubelet` rather than the deeper `/var/lib/kubelet/pods` because that makes Kubelet aware of the actual disk space available for ephemearl storage under the `/var/lib/kubelet` path. However, that causes the error we showed above.

I did try just symlinking `/var/lib/kubelet/pods` instead to more accurately mimick https://gist.github.com/bcressey/2d3e139dbf01a1e1562f5ba49118258b ... but we get a different and related error:

```
Warning  FailedMount  2s (x6 over 19s)  kubelet            MountVolume.SetUp failed for volume "pvc-d9fad636-812b-43cf-a412-6f30455726c8" : rpc error: code = Internal desc = Could not mount "fs-79dfd17e:/" at "/var/lib/kubelet/pods/4a9c5dbe-3b2d-4f32-8549-417d0b328e0a/volumes/kubernetes.io~csi/pvc-d9fad636-812b-43cf-a412-6f30455726c8/mount": mount failed: exit status 32
Mounting command: mount
Mounting arguments: -t efs -o accesspoint=fsap-08058070bbcb4c872,tls fs-79dfd17e:/ /var/lib/kubelet/pods/4a9c5dbe-3b2d-4f32-8549-417d0b328e0a/volumes/kubernetes.io~csi/pvc-d9fad636-812b-43cf-a412-6f30455726c8/mount
Output: Could not start amazon-efs-mount-watchdog, unrecognized init system "aws-efs-csi-dri"
b'mount.nfs4: mount point /var/lib/kubelet/pods/4a9c5dbe-3b2d-4f32-8549-417d0b328e0a/volumes/kubernetes.io~csi/pvc-d9fad636-812b-43cf-a412-6f30455726c8/mount does not exist'
```

**Comment by diranged** [comment]
If I recall correctly - the `bootstrap` containers run before `kubelet` is up and running, right? So if the `--root-dir` flag was somehow exposed through the Bottlerocket configuration interface, then we could create our `/mnt/containers` dir via our bootstrap container and then use the `api-client` to set the `--root-dir` flag for kubelet. Just a thought.

**Comment by diranged** [comment]
(Side note, I have opened case #10230969021 within our AWS account to track a long term solution to this problem)..

**Comment by bcressey** [comment]
My heartfelt apologies, @diranged - the 2 AM page is certainly not the sort of experience I want anyone to have.

I need to dig into the two errors. In the first case, it looks like there's an attempt to relabel the volume with the pod's private label (`system_u:object_r:data_t:s0:c577,c1017`). That's broken for two reasons: EFS doesn't support xattrs (so `security.selinux` can't be set), and that relabeling shouldn't happen for a `ReadWriteMany` volume.

The second case looks like an incompatibility between the symlink approach and CSI driver expectations, where the driver's mounts aren't being propagated back to the host.

**Comment by bcressey** [comment]
> I think it's time that Bottlerocket have an official way to manage storage before kubelet starts up... or that there are parameters exposed that let us reconfigure kubelet to point to a different storage path. Either way, the symlink option always felt like a hack... and the hack finally bit us.

I agree that this is the right long-term approach. The "ideal" bootstrap container does stuff that can't usefully be done at another time, and then makes API calls to put the host into the expected state. Doing stuff and then putting symlinks down to get the host to behave differently goes against the grain.

**Comment by bcressey** [comment]
The symlink for `/var/lib/kubelet/pods` approach looks to be incompatible with at least the EBS and EFS CSI plugins, if not most others. These pods mount `/var/lib/kubelet` with the "Bidirectional" (`rshared`) option, which is supposed to propagate new mounts on `/var` back to the host. When `kubelet` creates the mount point on the host, it's created under `/mnt`, which isn't a bind mount. When the CSI plugin is told to mount the volume there, it follows the symlink to its own `/mnt`, which doesn't propagate from the host, so the mount point doesn't exist. This causes the second error.

If we override kubelet's directories (`--root-dir`, `--volume-plugin-dir`) to point to the real paths under `/mnt`, and make `/var/lib/kubelet` a symlink, that causes a different problem. The `/mnt` path gets bound into `/var/lib/kubelet` in the CSI plugin container, which is great. The CSI plugin sees the mount point that `kubelet` creates, and continues. However, the plugin will create mounts under its own `/mnt`, which doesn't propagate back to the host. Everything proceeds without error, but pods will silently use local storage for everything, because the actual volume mount is only visible to the CSI plugin.

That leaves your first approach - relocate the `/var/lib/kubelet` directory. It looks like it could either be a symlink or a directory or a bind mount, but in any case it's important for compatibility that the entire `kubelet` tree resides on the same mount point.

The specific failure that's happening with the EFS CSI driver is this.

```
# relocated storage, broken
Jun 15 23:16:38 ip-10-0-80-6.us-west-2.compute.internal containerd[1886]: time="2022-06-15T23:16:38.179281028Z" level=info msg="CRI mounts =>
  [\"&Mount{ContainerPath:/efs,HostPath:/var/lib/kubelet/pods/aed38b69-e796-44a8-b45b-9cbde7278779/volumes/kubernetes.io~csi/pvc-d44d6520-8e8a-4618-9959c4514c3c7e06/mount,
    Readonly:false,SelinuxRelabel:true,Propagation:PROPAGATION_PRIVATE,}\" 

# default storage, working
Jun 15 23:41:03 ip-10-0-85-174.us-west-2.compute.internal containerd[1665]: time="2022-06-15T23:41:03.864036944Z" level=info msg="CRI mounts =>
  [\"&Mount{ContainerPath:/efs,HostPath:/var/lib/kubelet/pods/e78e12cd-e83b-4131-a364-2ecbe9188601/volumes/kubernetes.io~csi/pvc-4d030859-05a8-4375-9e58-c01127c3012a/mount,
    Readonly:false,SelinuxRelabel:false,Propagation:PROPAGATION_PRIVATE,}\" 
```

For some reason, either `kubelet` or the CSI plugin decides to set `SelinuxRelabel:true` in the relocated storage case, but not otherwise. That causes the first error when `containerd-cri` attempts to carry out the requested relabel, which fails because that's not supported by EFS.

I'm still digging but I'm getting closer.

**Comment by bcressey** [comment]
What happens is that `kubelet` [checks CSI mounts for SELinux support](https://github.com/kubernetes/kubernetes/blob/4b024fc4eeb4a3eeb831e7fddec52b83d0b072df/pkg/volume/csi/csi_mounter.go#L273) which [inspects mount info for the seclabel option](https://github.com/kubernetes/kubernetes/blob/4b024fc4eeb4a3eeb831e7fddec52b83d0b072df/pkg/volume/util/hostutil/hostutil_linux.go#L237), using the mount info from [the first prefix match of the path](https://github.com/kubernetes/kubernetes/blob/4b024fc4eeb4a3eeb831e7fddec52b83d0b072df/pkg/volume/util/hostutil/hostutil_linux.go#L191).

With no symlink in place, this will match a line in `/proc/self/mountinfo` like:
```
3780 275 0:648 / /var/lib/kubelet/pods/7912e91d-8977-4fc7-b689-c85550070726/volumes/kubernetes.io~csi/pvc-ba4a77bb-550e-4a60-908e-b5647d9a439d/mount rw,relatime shared:740 - nfs4 127.0.0.1:/ rw,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,noresvport,proto=tcp,port=20270,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1
```

(Note the lack of the "seclabel" option.)

However, with a symlink, the actual mount in `/proc/self/mountinfo` is the real host path:
```
3780 275 0:648 / /local/mnt/scratch/kubelet/pods/7912e91d-8977-4fc7-b689-c85550070726/volumes/kubernetes.io~csi/pvc-ba4a77bb-550e-4a60-908e-b5647d9a439d/mount rw,relatime shared:740 - nfs4 127.0.0.1:/ rw,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,noresvport,proto=tcp,port=20270,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1
```

This doesn't match the expected `/var/lib/kubelet/...` prefix, so the search continues until eventually it winds up at the original `/var` mount:
```
148 21 259:15 /var /var rw,nosuid,nodev,noexec,noatime - ext4 /dev/nvme1n1p1 rw,seclabel
```

`/var` is on local storage and does support `seclabel`, so the CRI mount gets `SELinuxLabel:true` set.

**Comment by bcressey** [comment]
That leaves us with two requirements for CSI drivers to work:
* `/var/lib/kubelet` must all reside on the same mount point (ignoring volume mounts, which come later)
* `/var/lib/kubelet` must be the mount point on the host, not a symlink to somewhere else

I'll check but I don't think it's possible to satisfy both requirements with bootstrap containers today. Binding the directory in `/mnt` to the directory in `/var` should work, but we don't propagate mounts in `/var` back to the host, only `/mnt`.

**Comment by diranged** [comment]
Thanks for that really thorough deep dive! I look forward to a solution...

**Comment by bcressey** [comment]
The only short-term fix to make CSI drivers work that I can see is to allow bootstrap containers to bind mount onto the storage directories, and have those mounts propagate back to the host.

This is conceptually simple but winds up being a larger change, in part because I'm also trying to eliminate the "redundant" entries in the mount list. Otherwise each volume mount will show up three times:
* `/var/lib/kubelet/...`
* `/local/var/lib/kubelet/...`
* `/mnt/storage/kubelet/...`

This will eventually cause a performance issue somewhere given that `systemd` and `kubelet` end up processing the full list on a regular basis.

I've mostly got everything sorted out so if testing continues to go well then I should have a PR up in the next few days. 

Unfortunately the switch to bind mounts also means that the previous approach of using symlinks simply won't work - the symlinks would interfere with passing the storage directories into the bootstrap container for bind mounting. I won't be sad to see the symlinks going away, but it means I also need to write sample code for having the same script work for old and new versions of Bottlerocket.

**Comment by bcressey** [comment]
I still need to work up an example of how to deal with "old" and "new" Bottlerocket approaches to mounting container storage in the same bootstrap container script.

**Comment by bcressey** [comment]
I've been mulling over the long term fix for a while. Right now the tentative design I have in mind is to extend the user-data / first-boot configuration interface to support "actions" as well as "settings".

```
[settings.updates]
ignore-waves = true
version-lock = "latest"

[actions.update]
commands = [
  [ "apply", "--check", "--reboot" ],
]
mode = "once"

[actions.ephemeral-storage]
commands = [
  [ "mount", "--format", "--bind" ],
]
mode = "always"
```

Where the "action" would be constrained to one of the sub-commands offered by `apiclient`. Those commands would in turn be backed by the actions API, so `apiserver` would have new methods like:
```
/actions/format-ephemeral-storage?block_device_type=lvm2;filesystem_type=ext4
/actions/mount-ephemeral-storage
/actions/bind-ephemeral-storage?target=kubelet;target=containerd
```

`apiclient` would have new sub-commands like:
```
apiclient ephemeral-storage format --block-device-type=lvm2 --filesystem-type=ext4
apiclient ephemeral-storage mount
apiclient ephemeral-storage bind --target=kubelet --target=containerd
```

Architecturally the goals would be to:
* identify common uses for bootstrap containers
* distill them into first-class API methods
* provide `apiclient` helper commands for them
* make it easy to invoke `apiclient` at the right moment

These "actions" would run at the same point in boot as bootstrap containers - after initial configuration, before the orchestrator services start. They'd support similar fields - being marked as essential, running on every boot vs. just once. They'd be less flexible, but also less of a chore to deal with.

**Comment by arnaldo2792** [comment]
Is the idea to support both bootstrap-containers and API "actions"?

If we support the `ephemeral-storage` action, and subcommands (`format`, `mount`, `bind`), doesn't that mean that we need to support in the host more tools to let the customers format their disks however they want? Aren't we concerned about the implications that has in the OS footprint? With bootstrap containers that isn't a concern (the container should have the tools required to format the disks)

```sh
apiclient ephemeral-storage format
```

Will that run right away? Or is that going to "schedule" the change, so that in the next boot the change is applied? I assume in the implementation we won't want `apiclient` or `apiserver` be the privileged processes performing the actions but rather `apiserver` will restart the appropriate systemd service in charge of performing the privileged actions (or something along those lines).

**Comment by diranged** [comment]
So on the one hand, I think that this ultimately is simpler for users.. so that's good. On the other hand, I have some questions.

1. How do you order the actions?
2. Do actions happen before or after bootstrap containers? Or both?
3. This means that bottlerocket is defining what `ephemeral-storage format` means ...
    a. are you going to raid-0 the volumes with mdadm? 
    b. what about using lvm?
    c. what if i want to control block size, or use mirroring?

The thing I likd about the bootstrap containers model was that it's up to me to decide how to handle the ephemeral storage volumes. The only thing I feel like I am missing is the ability to modify the kubelet configuration from the bootstrap containers.

Have you thought about just letting us mutate the kubelet config _to our hearts content_ from within a bootstrap container?

**Comment by bcressey** [comment]
@arnaldo2792 bootstrap containers would continue to be supported alongside actions. There'd need to be more host tools, like mdadm, lvm2, and devicemapper, but some of them might be required eventually anyhow in order to support scenarios like encrypting the local or private filesystems with a TPM-backed key.

@diranged --
> 1. How do you order the actions?

Good question. One approach might be to remove the "key must be a valid subcommand" restriction and let valid actions be grouped together:
```
[[actions]]
commands = [
  [ "set", "updates.ignore-waves=true" ],
  [ "update", "apply", "--check" ],
  [ "set", "updates.ignore-waves=false" ],
  [ "reboot" ],
]
mode = "once"
```

> 2. Do actions happen before or after bootstrap containers? Or both?

I had "after" in mind, so that all the potential desired settings would be in place. But "before" could make sense if the bootstrap container wanted to do something with the configured storage. Adding an early or late phase to support both, and defaulting to late, might be most flexible.

> 3. This means that bottlerocket is defining what `ephemeral-storage format` means ...
>    a. are you going to raid-0 the volumes with mdadm?
>    b. what about using lvm?
>    c. what if i want to control block size, or use mirroring?

Yeah, that's definitely the downside. I don't really love picking these values or making the API super flexible to parameterize everything. I'd tend to want a limited number of modes that were good enough for common cases, with bootstrap containers left for anything that needed more fine-tuning.

> The thing I liked about the bootstrap containers model was that it's up to me to decide how to handle the ephemeral storage volumes.

That is also the thing I like best about bootstrap containers!

**Comment by bcressey** [comment]
@diranged -
> Have you thought about just letting us mutate the kubelet config to our hearts content from within a bootstrap container?

I have thought a lot about this. It's one of the top questions that I get 😀 

The major pain point is that bootstrap containers only run once, but the kubelet config can be mutated at any time by modifying the underlying settings, which would clobber the changes from the bootstrap container.

I'm sort of bouncing between two ideas:
1. add a new container construct ("config containers") that can hook onto configuration files, and get a chance to modify them whenever they change.
2. add a "freeze" mechanism, where changes made to settings are not immediately applied but only on the next boot, which would give bootstrap containers a chance to modify them after they're rendered.

Neither seems great to me. The first would require a somewhat arcane, single purpose container. The second would let mutations happen in a natural way - appending to a file, rewriting with a `jq` pipeline, etc - but might break expectations elsewhere.

**Comment by bcressey** [comment]
An alternative that would make "config containers" much easier to use would be to just mount all of `/etc` into the container and give them the `super_t` label to let them modify anything there:
```
[settings.config-containers.my-container]
enabled = [ true | false ]
source = "..."
# whether to continue running restart commands if the container exits non-zero
essential = [ true | false ]
```

This solves one of the API concerns I had - that configuration files might be refactored or renamed across updates, which would break any file-specific hooks. Potentially it'd need to be combined with a smarter rollback mechanism, so if any essential bootstrap or config containers failed after an update, the system would flip back to the previous version and reboot.

It'd be a pretty radical change but also somewhat symmetrical in the overall architecture:
* "actions" for predefined, tested command sequences
* bootstrap containers as an escape hatch for other command sequences
* "settings" for predefined, tested configuration changes
* config containers as an escape hatch for other configuration changes

Using either escape hatch would introduce risk for updates, which would be backstopped by automatic rollback and require administrator intervention.

**Comment by bcressey** [comment]
I've updated #1991 with an example script for configuring ephemeral storage that works on 1.8.0 and the forthcoming 1.9.0. Obviously 1.8.0 will still have the problem with CSI drivers that led to this issue, but that should be fixed for 1.9.0.

**Comment by bcressey** [comment]
Resolving this since the fix is out in 1.9.0. I've opened #2343 to track the design for the config containers feature. I'll leave the actions API design dormant for now, since it's less clear to me that it'd be worthwhile.
